File: frontend/src/services/order-service.ts
import api from "@/lib/api";
import { API_PATHS } from "@/lib/constants";
import type { ApiResponse, PageResponse, PageParams } from "@/types/api";
import type {
  Order,
  OrderCreateRequest,
  OrderStatusUpdateRequest,
} from "@/types/order";

/**
 * Creates a new order.
 */
export async function createOrder(
  data: OrderCreateRequest
): Promise<ApiResponse<Order>> {
  const response = await api.post<ApiResponse<Order>>(
    API_PATHS.ORDERS.CREATE,
    data
  );
  return response.data;
}

/**
 * Gets the user's order list.
 */
export async function getUserOrders(
  page = 0,
  size = 12
): Promise<ApiResponse<PageResponse<Order>>> {
  const response = await api.get<ApiResponse<PageResponse<Order>>>(
    API_PATHS.ORDERS.LIST,
    { params: { page, size } }
  );
  return response.data;
}

/**
 * Gets an order by ID.
 */
export async function getOrderById(id: number): Promise<ApiResponse<Order>> {
  const response = await api.get<ApiResponse<Order>>(
    API_PATHS.ORDERS.DETAIL(id)
  );
  return response.data;
}

/**
 * Cancels an order.
 */
export async function cancelOrder(id: number): Promise<ApiResponse<Order>> {
  const response = await api.put<ApiResponse<Order>>(
    API_PATHS.ORDERS.CANCEL(id)
  );
  return response.data;
}

/**
 * Gets all orders (admin).
 */
export async function getAllOrders(
  page = 0,
  size = 12
): Promise<ApiResponse<PageResponse<Order>>> {
  const response = await api.get<ApiResponse<PageResponse<Order>>>(
    API_PATHS.ADMIN.ORDERS,
    { params: { page, size } }
  );
  return response.data;
}

/**
 * Updates an order's status (admin).
 */
export async function updateOrderStatus(
  id: number,
  data: OrderStatusUpdateRequest
): Promise<ApiResponse<Order>> {
  const response = await api.put<ApiResponse<Order>>(
    API_PATHS.ADMIN.ORDER_STATUS(id),
    data
  );
  return response.data;
}

File: backend/src/main/java/com/headless/ecommerce/service/OrderService.java
package com.headless.ecommerce.service;

import com.headless.ecommerce.dto.request.OrderCreateRequest;
import com.headless.ecommerce.dto.response.OrderResponse;
import com.headless.ecommerce.dto.response.PageResponse;
import com.headless.ecommerce.model.Order;
import com.headless.ecommerce.model.enums.OrderStatus;

/**
 * Order service interface for order management.
 */
public interface OrderService {

    /**
     * Creates a new order from the user's cart.
     *
     * @param userId the user ID
     * @param request the order creation request
     * @return the created order response
     */
    OrderResponse createOrder(Long userId, OrderCreateRequest request);

    /**
     * Gets a paginated list of orders for a user.
     *
     * @param userId the user ID
     * @param page page number (0-based)
     * @param size page size
     * @return paginated order response
     */
    PageResponse<OrderResponse> getUserOrders(Long userId, int page, int size);

    /**
     * Gets an order by ID.
     *
     * @param orderId the order ID
     * @param userId the user ID (for ownership verification)
     * @return the order response
     */
    OrderResponse getOrderById(Long orderId, Long userId);

    /**
     * Cancels an order.
     *
     * @param orderId the order ID
     * @param userId the user ID
     * @return the cancelled order response
     */
    OrderResponse cancelOrder(Long orderId, Long userId);

    /**
     * Updates the status of an order (admin operation).
     *
     * @param orderId the order ID
     * @param newStatus the new order status
     * @return the updated order response
     */
    OrderResponse updateOrderStatus(Long orderId, OrderStatus newStatus);

    /**
     * Gets a paginated list of all orders (admin).
     *
     * @param page page number (0-based)
     * @param size page size
     * @return paginated order response
     */
    PageResponse<OrderResponse> getAllOrders(int page, int size);

    /**
     * Restores stock for all items in a cancelled order.
     * Used by OrderTimeoutScheduler to avoid code duplication.
     *
     * @param order the cancelled order
     */
    void restoreStock(Order order);
}
